components/layout/query.rs

```diff
@@ -37,16 +40,111 @@ use style::values::specified::GenericGridTemplateComponent;
+    }
+
+    /// Get a scroll node that would represents this [ServoLayoutNode]'s transform and
+    /// calculate its transfrom from its root scroll node to the scroll node.
+    fn layout_node_to_root_transform(&self, node: ServoLayoutNode<'_>) -> Option<LayoutTransform> {
+        let scroll_tree_node_id = self.get_scroll_node_id_for_query(node)?;
+
+        Some(self.scroll_node_to_root_transform(&scroll_tree_node_id))
+    }
+
+    /// Traverse a scroll node to its root to calculate the transform.
+    // TODO(stevennovaryo): Add caching mechanism for this.
+    fn scroll_node_to_root_transform(&self, node_id: &ScrollTreeNodeId) -> LayoutTransform {
+        if let Some(scroll_tree) = self.scroll_tree {
+            let cur_node = scroll_tree.get_node(node_id);
+
+            // FIXME(stevennovaryo): Ideally we should optimize the computation of simpler
+            //                       transformation like translate as it could be done
+            //                       in smaller amount of operation compared to a normal
+            //                       matrix multiplication.
+            let node_transform = match &cur_node.info {
+                // To apply a transformation we need to make sure the rectangle's
+                // coordinate space is the same as reference frame's coordinate space.
+                // TODO(stevennovaryo): we are ignoring scaling (or zoom) consideration
+                //                      in transforming the coordinate space, since we are
+                //                      yet to implement it anyway.
+                SpatialTreeNodeInfo::ReferenceFrame(info) => info.transform.change_basis(
+                    info.frame_origin_for_query.x,
+                    info.frame_origin_for_query.y,
+                    0.0,
+                ),
+                SpatialTreeNodeInfo::Scroll(info) => {
+                    Transform3D::translation(-info.offset.x, -info.offset.y, 0.0)
+                },
+                SpatialTreeNodeInfo::Sticky(_) => Default::default(),
+            };
+
+            match cur_node.parent {
+                // If a node is not a root, accumulate the transforms.
+                Some(parent_id) => {
+                    let ancestors_transform = self.scroll_node_to_root_transform(&parent_id);
+                    node_transform.then(&ancestors_transform)
+                },
+                None => node_transform,
+            }
+        } else {
+            Default::default()
+        }
+    }
+
+    fn apply_post_composite_transform(
+        rect_to_transform: Rect<Au>,
+        transform: LayoutTransform,
+    ) -> Rect<Au> {
+        let layout_rect_union = au_rect_to_f32_rect(rect_to_transform).cast_unit();
+
+        if let Some(transformed_rect) = transform.outer_transformed_rect(&layout_rect_union) {
+            f32_rect_to_au_rect(transformed_rect.to_untyped())
+        } else {
+            rect_to_transform
+        }
+    }
```

Comment:
```suggestion
        transform.outer_transformed_rect(&layout_rect_union)
            .map(|transformed_rect| 32_rect_to_au_rect(transformed_rect.to_untyped())
            .unwrap_or(rect_to_transform)
    }
```

And in fact, I would simply move these three lines to where they are used twice. I think the function isn't adding too much here.